Update list of deprecations and removals to include missing parts. (#1533)
neo4j/neo4j-documentation: cypher/cypher-docs/src/docs/dev/deprecations-additions-and-compatibility.asciidoc

[source, cypher, role="noheader"]
----
SHOW {POINT \| RANGE \| TEXT} INDEXES

----
a|
label:functionality[]
label:removed[]
[source, cypher, role="noheader"]
----

> USING BTREE INDEXES

Reviewer: `USING BTREE INDEX`

----
WHERE NOT isEmpty([1, 2, 3])
----

a|
label:functionality[]
label:removed[]
[source, cypher, role="noheader"]
----
distance(n.prop, point({x:0, y:0})
----
a|
Replaced by:
[source, cypher, role="noheader"]
----
point.distance(n.prop, point({x:0, y:0})
----

a|
label:functionality[]
label:removed[]
[source, cypher, role="noheader"]
----
point({x:0, y:0}) <= point({x:1, y:1}) <= point({x:2, y:2})
----
a|
Using inequality operators `<`, `<=`, `>`, and `>=` on spatial points is removed.
Please instead use:
[source, cypher, role="noheader"]
----
point.withinBBox(point({x:1, y:1}), point({x:0, y:0}), point({x:2, y:2}))
----

a|
label:syntax[]
label:removed[]
[source, cypher, role="noheader"]
----
USING PERIODIC COMMIT ...
----
a|
Replaced by:
[source, cypher, role="noheader"]
----
CALL {
  ...
} IN TRANSACTIONS
----

a|
label:syntax[]
label:removed[]
[source, cypher, role="noheader"]
----
CREATE (a {prop:7})-[r:R]->(b {prop: a.prop})
----
a|

> Patterns where a variable introduced in the pattern is also referenced from the same pattern are no longer supported.

Reviewer 2: This should probably say that it is specific to `CREATE`

Author: Yes good catch

a|
label:syntax[]
label:removed[]
[source, cypher, role="noheader"]
----
CALL { RETURN 1 }
----
a|
Unaliased expressions are no longer supported in subquery `RETURN` clauses. Replaced by:
[source, cypher, role="noheader"]
----
CALL { RETURN 1 AS one }
----

a|
label:syntax[]
label:removed[]
[source, cypher, role="noheader"]
----
MATCH (a) RETURN (a)--()
----
a|

> Pattern expressions for checking existence are no longer supported, except for use in `WHERE` clauses. Use a pattern comprehension instead. Replaced by:

Reviewer 2: ```suggestion
Pattern expressions producing lists of paths are no longer supported. Use a pattern comprehension instead. (Pattern expressions can still be used as existence predicates in `WHERE` clauses, and in `exists()`). Replaced by:
```

Author: I intentionally left out `and in exists()` from this table, as it is for 5.0 where `exists()` has been removed. But I will update it to something similar as your suggestion

Reviewer 3: I think you have a point.

You could consider writing something like
> as existence predicates, **for example** in `WHERE` clauses.

as pattern expressions can also be used in other boolean contexts such as `CASE` statements.

Reviewer 3: Also, from an English point of view, I would expect either

> Instead, use **a** pattern comprehension:

or

> Instead, use pattern comprehension**s**:

Same applies to the section deprecated in 4.3.

Reviewer 4: You are absolutely right. I just left the way it was because I thought it could be used as a more generic technical term, but I guess adding an "a" will make the sentence better. Thanks for flagging this!

----
WHERE NOT isEmpty([1, 2, 3])
----

a|
label:functionality[]
label:removed[]
[source, cypher, role="noheader"]
----
distance(n.prop, point({x:0, y:0})
----
a|
Replaced by:
[source, cypher, role="noheader"]
----
point.distance(n.prop, point({x:0, y:0})
----

a|
label:functionality[]
label:removed[]
[source, cypher, role="noheader"]
----
point({x:0, y:0}) <= point({x:1, y:1}) <= point({x:2, y:2})
----
a|

> Using inequality operators `<`, `<=`, `>`, and `>=` on spatial points is removed.

Reviewer 4: ```suggestion
The ability to use the inequality operators `<`, `<=`, `>`, and `>=` on spatial points was removed.
```

Author: It looks like all other places in this page are using 'is/are removed' and 'is/are deprecated' rather than 'was' so if you want to change that, it would be better to do it as a separate task for the whole page. I will change the first part though

> Please instead use:

Reviewer 4: ```suggestion
Instead, use:
```

[source, cypher, role="noheader"]
----
point.withinBBox(point({x:1, y:1}), point({x:0, y:0}), point({x:2, y:2}))
----

a|
label:syntax[]
label:removed[]
[source, cypher, role="noheader"]
----
USING PERIODIC COMMIT ...
----
a|
Replaced by:
[source, cypher, role="noheader"]
----
CALL {
  ...
} IN TRANSACTIONS
----

a|
label:syntax[]
label:removed[]
[source, cypher, role="noheader"]
----
CREATE (a {prop:7})-[r:R]->(b {prop: a.prop})
----
a|

> `CREATE` clauses where a variable introduced in the pattern is also referenced from the same pattern are no longer supported.

Reviewer 4: This sentence is a bit confusing and too long, but I'm not sure if my suggestion is correct, so please take a look:

```suggestion
It is no longer allowed to have `CREATE` clauses in which a variable introduced in the pattern is also referenced from the same pattern.
```

a|
label:syntax[]
label:removed[]
[source, cypher, role="noheader"]
----
CALL { RETURN 1 }
----
a|
Unaliased expressions are no longer supported in subquery `RETURN` clauses. Replaced by:
[source, cypher, role="noheader"]
----
CALL { RETURN 1 AS one }
----

a|
label:syntax[]
label:removed[]
[source, cypher, role="noheader"]
----
MATCH (a) RETURN (a)--()
----
a|

> Pattern expressions producing lists of paths are no longer supported. Use a pattern comprehension instead. Pattern expressions can still be used as existence predicates in `WHERE` clauses. Replaced by:

Reviewer 4: Again, not sure if I changed the meaning, so please take a look:

```suggestion
Pattern expressions producing lists of paths are no longer supported, but they can still be used as existence predicates in `WHERE` clauses.
Instead, use pattern comprehension:
```

[source, cypher, role="noheader"]
----
MATCH (a) RETURN [p=(a)--() \| p]
----

a|
label:functionality[]
label:removed[]
[source, cypher, role="noheader"]
----
MATCH (n) RETURN n.propertyName_1, n.propertyName_2 + count(*)
----
a|
Implied grouping keys are no longer supported.

> Only expressions that do _not_ contain aggregations are grouping keys.

Reviewer 4: ```suggestion
Only expressions that do _not_ contain aggregations are still considered grouping keys.
```

> In expressions that contain aggregations, the leaves of expressions must be one of:

Reviewer 4: ```suggestion
In expressions that contain aggregations, the leaves must be either:
```

- An aggregation
- A literal

> - A Parameter

Reviewer 4: ```suggestion
- A parameter
```

> - A variable - *ONLY IF* that variable is also one of: +

Reviewer 4: ```suggestion
- A variable, *ONLY IF* it is either:
```

1) A projection expression on its own (e.g. the `n` in `RETURN n AS myNode, n.value + count(*)`) +
2) A local variable in the expression (e.g the `x` in `RETURN n, n.prop + size([ x IN range(1, 10) \| x ]`)

> - Property access - *ONLY IF* that property access is also a projection expression on its own (e.g. the `n.prop` in `RETURN n.prop, n.prop + count(*)`) +

Reviewer 4: ```suggestion
- Property access, *ONLY IF* it is also a projection expression on its own (e.g. the `n.prop` in `RETURN n.prop, n.prop + count(*)`) +
```

> - Map access - *ONLY IF* that map access is also a projection expression on its own (e.g. the `map.prop` in `WITH {prop: 2} AS map RETURN map.prop, map.prop + count(*)`)

Reviewer 4: ```suggestion
- Map access, *ONLY IF* it is also a projection expression on its own (e.g. the `map.prop` in `WITH {prop: 2} AS map RETURN map.prop, map.prop + count(*)`)
```

----
WHERE NOT isEmpty([1, 2, 3])
----

a|
label:functionality[]
label:removed[]
[source, cypher, role="noheader"]
----
distance(n.prop, point({x:0, y:0})
----
a|
Replaced by:
[source, cypher, role="noheader"]
----
point.distance(n.prop, point({x:0, y:0})
----

a|
label:functionality[]
label:removed[]
[source, cypher, role="noheader"]
----
point({x:0, y:0}) <= point({x:1, y:1}) <= point({x:2, y:2})
----
a|

> The ability to use operators `<`, `<=`, `>`, and `>=` on spatial points is removed.

Reviewer: The ability to use operators `+<+`, `+<=+`, `+>+`, and `+>=+` on spatial points is removed.

Escape operators, especially `<=` or it will not show these characters in asciidoc.

----
a|
Implied grouping keys are deprecated.
Only expressions that do _not_ contain aggregations are still considered grouping keys.
In expressions that contain aggregations, the leaves must be either:

- An aggregation
- A literal
- A parameter
- A variable, *ONLY IF* it is either: +
1) A projection expression on its own (e.g. the `n` in `RETURN n AS myNode, n.value + count(*)`) +
2) A local variable in the expression (e.g the `x` in `RETURN n, n.prop + size([ x IN range(1, 10) \| x ]`)
- Property access, *ONLY IF* it is also a projection expression on its own (e.g. the `n.prop` in `RETURN n.prop, n.prop + count(*)`) +

> - Map access, *ONLY IF* it is also a projection expression on its own (e.g. the `map.prop` in `WITH {prop: 2} AS map RETURN map.prop, map.prop + count(*)`)

Reviewer: Was this cherry picked to `4.4` ???

Author: Yes, in https://github.com/neo4j/neo4j-documentation/pull/1558/files

} IN TRANSACTIONS
----

a|
label:syntax[]
label:deprecated[]
[source, cypher, role="noheader"]
----
CREATE (a {prop:7})-[r:R]->(b {prop: a.prop})
----
a|

> `CREATE` clauses where a variable introduced in the pattern is also referenced from the same pattern are deprecated.

Reviewer 4: Same as previous suggestion

```suggestion
It is no longer allowed to have `CREATE` clauses in which a variable introduced in the pattern is also referenced from the same pattern.
```

Author: This is in the table for 4.4, where the feature was deprecated but still allowed, so it should say deprecated. But I can try to reformulate it slightly still

} IN TRANSACTIONS
----

a|
label:syntax[]
label:deprecated[]
[source, cypher, role="noheader"]
----
CREATE (a {prop:7})-[r:R]->(b {prop: a.prop})
----
a|
`CREATE` clauses in which a variable introduced in the pattern is also referenced from the same pattern are deprecated.

> 

Reviewer: Was this cherry picked to `4.4` ???

Author: Yes, in https://github.com/neo4j/neo4j-documentation/pull/1558/files

SHOW {POINT \| RANGE \| TEXT} INDEXES
----

a|
label:functionality[]
label:deprecated[]
[source, cypher, role="noheader"]
----
USING BTREE INDEX
----
a|
B-tree indexes are deprecated.

Replaced by:
[source, cypher, role="noheader"]
----
USING {POINT \| RANGE \| TEXT} INDEX

> ----

Reviewer: Was this cherry picked to `4.4` ???

Author: Yes, in https://github.com/neo4j/neo4j-documentation/pull/1558/files

point({x:0, y:0}) <= point({x:1, y:1}) <= point({x:2, y:2})
----
a|
The ability to use the inequality operators `<`, `<=`, `>`, and `>=` on spatial points is deprecated.

> Instead, use:

Reviewer: Was this cherry picked to `4.4` ???

Author: Yes, in https://github.com/neo4j/neo4j-documentation/pull/1558/files

----
ON HOME GRAPH
----

a|
label:syntax[]
label:deprecated[]
[source, cypher, role="noheader"]
----
MATCH (a) RETURN (a)--()
----
a|

> Pattern expressions for checking existence are deprecated, except for use in `WHERE` clauses and in `exists()`. Use a pattern comprehension instead. Replaced by:

Reviewer 2: ```suggestion
Pattern expressions producing lists of paths are deprecated. Use a pattern comprehension instead. (Pattern expressions can still be used as existence predicates in `WHERE` clauses, and in `exists()`). Replaced by:
```

----
ON HOME GRAPH
----

a|
label:syntax[]
label:deprecated[]
[source, cypher, role="noheader"]
----
MATCH (a) RETURN (a)--()
----
a|

> Pattern expressions producing lists of paths are deprecated. Use a pattern comprehension instead. Pattern expressions can still be used as existence predicates in `WHERE` clauses and in `exists()`. Replaced by:

Reviewer 4: Same suggestion from above:

```suggestion
Pattern expressions producing lists of paths are no longer supported, but they can still be used as existence predicates in `WHERE` clauses.
Instead, use pattern comprehension:
```

Author: Similar to the one above, this is in the table for 4.3 so it should say deprecated but I will try to reformulate it to something more like this. As for the `exists()` part, the `exists()` function has been removed in 5.0, but it is still valid when talking about 4.3, so I think I prefer to keep it here

----
ON HOME GRAPH
----

a|
label:syntax[]
label:deprecated[]
[source, cypher, role="noheader"]
----
MATCH (a) RETURN (a)--()
----
a|
Pattern expressions producing lists of paths are deprecated, but they can still be used as existence predicates, for example in `WHERE` clauses.
Instead, use a pattern comprehension:
[source, cypher, role="noheader"]
----
MATCH (a) RETURN [p=(a)--() \| p]

> ----

Reviewer: Was this cherry picked to `4.3` ???
Was this cherry picked to `4.4` ???

Author: Yes in https://github.com/neo4j/neo4j-documentation/pull/1558/files and https://github.com/neo4j/neo4j-documentation/pull/1559/files

----
SHOW CONSTRAINTS YIELD *
----

a|
label:syntax[]
label:deprecated[]
[source, cypher, role="noheader"]
----
CALL { RETURN 1 }
----
a|
Unaliased expressions are deprecated in subquery `RETURN` clauses. Replaced by:
[source, cypher, role="noheader"]
----
CALL { RETURN 1 AS one }

> ----

Reviewer: Was this cherry picked to `4.2` ???
Was this cherry picked to `4.3` ???
Was this cherry picked to `4.4` ???

Author: I backported it to 4.3 and 4.4, but not 4.2 as that is not supported anymore
